x-pack/plugin/slm/src/internalClusterTest/java/org/elasticsearch/xpack/slm/SLMStatDisruptionIT.java

```diff
@@ -0,0 +1,550 @@
+
+        createRandomIndex(idxName, dataNode);
+        createRepository(repoName, "mock");
+        createSnapshotPolicy(policyName, "snap", NEVER_EXECUTE_CRON_SCHEDULE, repoName, idxName);
+
+        ensureGreen();
+
+        String snapshotName = executePolicy(masterNode, policyName);
+        logger.info("Created snapshot: " + snapshotName);
+
+        waitForSnapshot(repoName, snapshotName);
+
+        assertBusy(() -> {
+            assertSnapshotSuccess("test-repo", snapshotName);
+            logger.info("Verified that snapshot was not successful");
+        }, 1, TimeUnit.MINUTES);
+
+        assertBusy(() -> {
+            var snapshotLifecyclePolicyMetadata = getSnapshotLifecyclePolicyMetadata(policyName);
+            assertNull(snapshotLifecyclePolicyMetadata.getLastFailure());
+            assertNotNull(snapshotLifecyclePolicyMetadata.getLastSuccess());
+            assertEquals(snapshotLifecyclePolicyMetadata.getInvocationsSinceLastSuccess(), 0);
+            assertEquals(snapshotLifecyclePolicyMetadata.getPreRegisteredRuns(), 0);
+            logger.info("Verified stats: invocationsSinceLastSuccess = 0, preRegisteredRuns = 0");
+        }, 1, TimeUnit.MINUTES);
+    }
+
+    /**
+     * Test that after a failure then a success, preRegisteredRuns from failure is added to invocationsSinceLastSuccess.
+     */
+    public void testFailSnapshotFailStatsThenSuccessRecoverStats() throws Exception {
+        final String idxName = "test-idx";
+        final String repoName = "test-repo";
+        final String policyName = "test-policy";
+
+        internalCluster().startMasterOnlyNodes(1);
+        final String masterNode = internalCluster().getMasterName();
+        final String dataNode = internalCluster().startDataOnlyNode();
+        ensureStableCluster(2);
+
+        NetworkDisruption networkDisruption = isolateMasterDisruption(NetworkDisruption.DISCONNECT);
+        internalCluster().setDisruptionScheme(networkDisruption);
+
+        // Listener that stops disrupting network only after snapshot completion
+        CountDownLatch latch = new CountDownLatch(1);
+        internalCluster().clusterService(masterNode)
+            .addListener(new WaitForSnapshotListener(repoName, networkDisruption, latch));
+
+        createRandomIndex(idxName, dataNode);
+        createRepository(repoName, "mock");
+        createSnapshotPolicy(policyName, "snap", NEVER_EXECUTE_CRON_SCHEDULE, repoName, idxName);
+
+        ensureGreen();
+
+        networkDisruption.startDisrupting();
+        String snapshotName = executePolicy(masterNode, policyName);
+        logger.info("Created snapshot: " + snapshotName);
+
+        // wait for snapshot to complete and network disruption to stop
+        assertTrue(latch.await(1, TimeUnit.MINUTES));
+
+        // restart master so failure stat is lost
+        // TODO this relies on a race condition. The node restart must happen before stats are stored in cluster state, but this is not guaranteed.
```

Comment:
There is a race condition here that I haven't figured how to remove. If the failure is recorded in the cluster state: https://github.com/elastic/elasticsearch/blob/ff7f00e553ccc0f9786a78d55b839b46b0e9ef03/x-pack/plugin/slm/src/main/java/org/elasticsearch/xpack/slm/SnapshotLifecycleTask.java#L131

before the master is restarted, the failure stat will not be lost. I'd like to test this without a race condition, but not sure how to do so.